File: input_processing/R/cfgather.R
sink("gamslog.txt",append=TRUE)

library(reshape2)
library(gdxrrw)

print("Beginning calculations in inputs\\cf\\cfgather.R")


#cfgather.R
#Args=list()
#Args[1]="D:\\Max\\r2_na\\"
#Args[2]="C:\\GAMS\\win64\\24.7\\"
#Args[3]="inputs\\gdxfiles\\harm.gdx"

colid = c("i","j","k","l","m","n","o","p")


if(!exists("Args")) Args=commandArgs(TRUE)

indir = paste(Args[1])
setwd(indir)

igdx(paste(Args[2]))
gdxfile = file.path("inputs","cf","cfdata.gdx")
distpvscen = as.character(Args[3])
outdir = paste0(Args[4])


rs = as.data.frame(read.csv(file.path(indir,"inputs","rsmap.csv")))
colnames(rs) = c("r","s")


####################
# Wind
####################

wind = as.data.frame(rgdx.param(gdxfile,"CF_Corr",names=c(colid[1:(length(colnames(rgdx.param(gdxfile,"CF_Corr")))-1)],"value")))
wind$c = paste(wind$k,"_",gsub("class","",wind$j),sep="")
wind$i = paste("s",wind$i,sep="")
colnames(wind) = c("s","cl","it","h","value","i")
wind=merge(wind,rs,by=c("s"))
wind = wind[,c("r","s","i","h","value")]

wcf = as.data.frame(rgdx.param(gdxfile,"CFW_all",names=c(colid[1:(length(colnames(rgdx.param(gdxfile,"CFW_all")))-1)],"value")))
wcf$i = paste(wcf$i,gsub("class","",wcf$k),sep="_")
wcf = wcf[,c("j","i","value")]
wcf = dcast(wcf,j~i,value.var="value")


####################
# UPV
####################

upv = as.data.frame(rgdx.param(gdxfile,"CFUPV",names=c(colid[1:(length(colnames(rgdx.param(gdxfile,"CFUPV")))-1)],"value")))
upv$c = paste("UPV","_",gsub("class","",upv$k),sep="")
upv$s = "sk"
colnames(upv) = c("r","h","cl","value","i","s")
upv = upv[,c("r","s","i","h","value")]

####################
# DUPV
####################

dupv = as.data.frame(rgdx.param(gdxfile,"CFDUPV",names=c(colid[1:(length(colnames(rgdx.param(gdxfile,"CFDUPV")))-1)],"value")))
dupv$c = paste("DUPV","_",gsub("class","",dupv$k),sep="")
dupv$s = "sk"
colnames(dupv) = c("r","h","cl","value","i","s")
dupv = dupv[,c("r","s","i","h","value")]

####################
# distPV
####################


distpv = as.data.frame(read.csv(file.path(indir,"inputs","dGen_Model_Inputs",distpvscen,paste0("distPVCF_", distpvscen, ".csv"))))
colnames(distpv)[1] = "ba"
distpv = melt(distpv,id=c("ba"))
colnames(distpv) = c("r","h","value")
distpv$i = "distPV"
distpv$s = "sk"

distpv = distpv[,c("r","s","i","h","value")]


####################
# CSP
####################

# CFCspwStorallyears in heritage ReEDS is the adjusted CSP-ws field capacity factor from CFCSPws_tower;
# data for tower system starts from 2018 and are the same across years (no need for a year dimension)
# Note that in the most version, CFCSPws_tower (read from field_capacity_factor.csv file) has already incorporated the availability factors
# (CFCSPws_adj_factor_tower and CSPws_avail_factor)
# so can directly read from CFCSPws_tower

# CF for CSP is from SAM modeling results, and assumes SM=1 (i.e. the numbers will need to be multiplied by the actual SM of certain configurations)

csp = as.data.frame(rgdx.param(gdxfile,"CFCSPws_tower",names=c(colid[1:(length(colnames(rgdx.param(gdxfile,"CFCSPws_tower")))-1)],"value")))
colnames(csp) = c("s","h","i","value")
csp$i = gsub("class","_",csp$i)
csp$i = gsub("csp","csp1",csp$i)
csp$s <- paste0("s", csp$s)
csp <- csp[,c("s","h","i","value")]
csp=merge(csp,rs,by=c("s"))

csp = dcast(csp,r+s+i~h,value.var="value")
csp[is.na(csp)] <- 0
csp$H5 <- 0
csp$H9 <- 0
csp <- csp[,c("r","s","i",sprintf("H%s",seq(1:17)))]

csp_temp <- csp
csp_temp$i <- gsub("csp1","csp2",csp_temp$i)
csp <- rbind(csp, csp_temp)
rm(csp_temp)

###################
#output writing
###################




allcf = rbind(wind,upv,dupv,distpv)
allcf = dcast(allcf,r+s+i~h,value.var="value")
allcf[is.na(allcf)] = 0
allcf <- rbind(allcf, csp)

##############################
#hydro mingen
###############################

hydro = as.data.frame(rgdx.param(gdxfile,"minplantload_hy",names=c(colid[1:(length(colnames(rgdx.param(gdxfile,"minplantload_hy")))-1)],"value")))
hydro$j = substr(hydro$j,1,4)
colnames(hydro) = c("i","szn","r","value")
hydro = dcast(hydro,i+r~szn,value.var="value")


#write.table(out,"cfout.txt",row.names=FALSE,col.names=FALSE,quote=FALSE)
print(paste("writing capacity factor data to:",outdir))
write.csv(allcf,paste0(outdir,"cfout.csv"),row.names=FALSE,quote=FALSE)
# write.csv(csp,paste0(outdir,"cfout_csp.csv"),row.names=FALSE,quote=FALSE)
write.csv(wcf,paste0(outdir,"windcfout.csv"),row.names=FALSE,quote=FALSE)
write.csv(hydro,paste0(outdir,"minhyd.csv"),row.names=FALSE,quote=FALSE)

File: input_processing/cfgather.py
#%% IMPORTS ###
import pandas as pd
import os
import argparse
# Time the operation of this script
from ticker import toc, makelog
import datetime
tic = datetime.datetime.now()

#%% Model Inputs
parser = argparse.ArgumentParser(description="""This file outputs capacity factor data""")

parser.add_argument("reeds_dir", help="ReEDS directory")
parser.add_argument("inputs_case", help="output directory")

args = parser.parse_args()
reeds_dir = args.reeds_dir
inputs_case = args.inputs_case

# #%% Inputs for testing
# reeds_dir = os.path.expanduser('~/github2/ReEDS-2.0/')
# inputs_case = os.path.join(
#     reeds_dir,'runs','v20221215_hourlycleanupM4_Pacific','inputs_case')

#%% Set up logger
log = makelog(scriptname=__file__, logpath=os.path.join(inputs_case,'..','gamslog.txt'))

#%% Inputs from switches
sw = pd.read_csv(
    os.path.join(inputs_case, 'switches.csv'), header=None, index_col=0, squeeze=True)

distpvscen = sw.distpvscen
GSw_IndividualSites = int(sw.GSw_IndividualSites)
GSw_SitingWindOns = sw.GSw_SitingWindOns
GSw_SitingWindOfs = sw.GSw_SitingWindOfs
GSw_SitingUPV = sw.GSw_SitingUPV
GSw_PVB = int(sw.GSw_PVB)
GSw_PVB_Types = (
    [int(i) for i in sw.GSw_PVB_Types.split('_')] if int(GSw_PVB)
    else []
)

#%% Procedure
print('Beginning calculations in cfgather.py')

# Read in set of valid regions
val_r = pd.read_csv(
        os.path.join(inputs_case, 'val_r.csv'), squeeze=True, header=None).tolist()
val_r_cap = pd.read_csv(
        os.path.join(inputs_case, 'val_r_cap.csv'), squeeze=True, header=None).tolist()

colid = ['i','j','k','l','m','n','o','p']

rs = pd.read_csv(os.path.join(inputs_case,'rsmap.csv')).rename(columns={'*r':'r'})
rs.columns = ['r','s']

rs = rs.loc[rs['r'].isin(val_r)]


###########
#Wind
###########
windons = pd.read_csv(
    os.path.join(
        reeds_dir, 'inputs', 'cf','wind-ons_cf_ts_{}-{}.csv'
    .format('site' if GSw_IndividualSites else 'sreg', GSw_SitingWindOns))
).drop('cfsigma', axis=1, errors='ignore').assign(type='wind-ons')


#%% Offshore wind
windofs = pd.read_csv(
    os.path.join(
        reeds_dir, 'inputs', 'cf','wind-ofs_cf_ts_{}-{}.csv'
    .format('site' if GSw_IndividualSites else 'sreg', GSw_SitingWindOfs))
).drop('cfsigma', axis=1, errors='ignore').assign(type='wind-ofs')

wind = pd.concat([windons,windofs])
wind['tech'] = wind['type'] + '_' + wind['class'].astype(str)
wind['class'] = 'class' + wind['class'].astype(str)
wind.columns = ['r','cl','h','value','it','i']
wind = wind[['r','i','h','value']]

##############
#UPV
##############

upv = (
    pd.read_csv(os.path.join(reeds_dir,'inputs','cf','upv_cf_ts-{}.csv'.format(GSw_SitingUPV)))
    .rename(columns={'region':'r','timeslice':'h','class':'i','cfmean':'value'})
    .drop('cfsigma', axis=1, errors='ignore')
)
upv.i = upv.i.map(lambda x: 'upv_{}'.format(x))
upv = upv.drop(upv.loc[upv.value==0].index).reset_index(drop=True)

#############
#DUPV
#############

dupv = pd.read_csv(os.path.join(reeds_dir,'inputs','cf','cf_dupv.csv'))

#############
#distpv
#############

distpv = pd.read_csv(
    os.path.join(reeds_dir,'inputs','dGen_Model_Inputs',distpvscen,'distPVCF_'+distpvscen+'.csv')
).rename(columns = {'Unnamed: 0':'r'})

distpv = distpv.melt('r')
distpv.columns = ['r','h','value']
distpv['i'] = 'distpv'
distpv = distpv[['r','i','h','value']]


#####################
#%% Hybrid PV+battery
#####################

### Get the PVB types
pvb_ilr = pd.read_csv(
    os.path.join(inputs_case, 'pvb_ilr.csv'),
    header=0, names=['pvb_type','ilr'], index_col='pvb_type', squeeze=True)

pvb = {}
for pvb_type in GSw_PVB_Types:
    ilr = int(pvb_ilr['pvb{}'.format(pvb_type)] * 100)
    ### UPV uses ILR = 1.3, so use its AC profile if ILR = 1.3
    acfile = 'upv' if ilr == 130 else 'upv{}'.format(ilr)
    ### Get timeslice AC output
    pvb_cf_ac = (
        pd.read_csv(os.path.join(reeds_dir,'inputs','cf','{}_cf_ts-{}.csv'.format(
            acfile, GSw_SitingUPV)))
        .rename(columns={'region':'r','timeslice':'h','class':'i','cfmean':'value'})
        .set_index(['r','i','h'])['value']
    )
    ### Get timeslice clipping (given as AC CF, i.e. after inverter losses are applied)
    pvb_cf_clipping = (
        pd.read_csv(os.path.join(reeds_dir,'inputs','cf','upv{}clip_cf_ts-{}.csv'.format(
            ### NOTE: We don't yet have clipping profiles for limited and open access
            ### from reV, so for now we use clipping from reference access for all PV
            ### siting scenarios. Once the open and limited clipping profiles are
            ### added, change 'reference' to GSw_SitingUPV in the next line.
            ilr, 'reference')))
        .rename(columns={'region':'r','timeslice':'h','class':'i','cfmean':'value'})
        .set_index(['r','i','h'])['value']
    )
    ### Sum the AC and clipping profiles and rename to match ReEDS usage
    pvb_cf = (
        pvb_cf_ac 
        + pvb_cf_clipping.to_frame().merge(
            pvb_cf_ac.to_frame()[[]],
            left_index=True, right_index=True,
            ### If there are any missing [r,i,h] values in the clipping file,
            ### set clipping to zero for those values
            how='right').fillna(0)['value']
    ).loc[pvb_cf_ac.index].reset_index()

    pvb_cf.i = pvb_cf.i.map(lambda x: 'PVB{}_{}'.format(pvb_type, x))
    pvb[pvb_type] = pvb_cf.drop(pvb_cf.loc[pvb_cf.value==0].index)

pvb = (
    pd.concat(pvb, axis=0, ignore_index=True) if GSw_PVB
    else pd.DataFrame(columns=['r','i','h','value']))

#############
#CSP
#############

#Read in tech-subset-table.csv to determine number of configureations
tst = pd.read_csv(os.path.join(reeds_dir,'inputs','tech-subset-table.csv'))
csp_configs = tst[(tst['CSP']=='YES') & (tst['STORAGE'] == 'YES')].shape[0]

# Note that in the most version, cf_csp_ws_tower (read from field_capacity_factor.csv file) has 
# already incorporated the availability factors (CFCSPws_adj_factor_tower and CSPws_avail_factor
# from Heritage ReEDS) so can directly read from cf_csp_ws_tower

# CF for CSP is from SAM modeling results, and assumes a solar multiple (SM) = 1.
# The solar multiple is applied in c_supplymodel.
csp_in = pd.read_csv(os.path.join(reeds_dir,'inputs','cf','cf_csp_ws_tower.csv'))

### Add profiles for each csp tech
csp = pd.concat([
    csp_in.assign(i=csp_in.i.str.replace('csp1',f'csp{i+1}')) for i in range(csp_configs)
], axis=0)

###### csp-ns
# Capacity factors for CSP-ns are developed using typical DNI year (TDY) hourly resource data
# (Habte et al. 2014) from 18 representative sites. The TDY weather files are processed through
# the CSP modules of SAM to develop performance characteristics for a system with a
# solar multiple of 1.4. These representative sites have an average DNI range of
# 7.25-7.5 kWh/m2/day (see "Class 3" in Table 4 of the ReEDS Model Documnetation: Version 2016).
# Habte, A., A. Lopez, M. Sengupta, and S. Wilcox. 2014. Temporal and Spatial Comparison of
# Gridded TMY, TDY, and TGY Data Sets. Golden, CO: National Renewable Energy Laboratory.
# http://www.osti.gov/scitech/biblio/1126297.
cf_cspns = pd.read_csv(
    os.path.join(reeds_dir,'inputs','cf','cf_cspns.csv'),
    header=None, names=['h','value'], index_col='h')
### Scale by availability
scalars = pd.read_csv(
    os.path.join(inputs_case,'scalars.csv'),
    header=None, names=['scalar','value','comment'], index_col='scalar')['value']
cf_cspns *= scalars['avail_cspns']
### All regions use the same CF for csp-ns
cspns = pd.concat({(r,'csp-ns'): cf_cspns for r in rs.s}, names=['r','i','h']).reset_index()

#######################
### Combine VRE CFs
#######################
allcf = pd.concat([wind,upv,dupv,distpv,pvb,csp,cspns])
allcf.h = allcf.h.str.lower()
allcf = (
    ## Filter for valid regions
    allcf.loc[allcf.r.isin(val_r_cap)]
    .rename(columns={'i':'*i','value':'cf'})
    [['*i','r','h','cf']]
).copy()

#######################
#Hydro mingen
#######################

hydro = pd.read_csv(os.path.join(reeds_dir,'inputs','cf','hydro_mingen.csv'))

# Filter for valid regions
hydro = hydro.loc[hydro['r'].isin(val_r)]

hydro = hydro.pivot_table(index =['i','r'],columns = 'szn',values = 'value').fillna(0).reset_index()
hydro = hydro.round(8)

#%% Save the outputs
print('writing capacity factor data to: '+inputs_case)
allcf.round(6).to_csv(os.path.join(inputs_case,'cf_vre.csv'), index=False)
hydro.to_csv(os.path.join(inputs_case,'hydmin.csv'),index=False)

toc(tic=tic, year=0, process='input_processing/cfgather.py', 
    path=os.path.join(inputs_case,'..'))
